### ISSUE TITLE
Tensor conversion requested dtype float32_ref for Tensor with dtype float32?

### ISSUE THREAD

**Issue by cjjjy** [tensorflow/tensorflow#24800]
word_embeddings = tf.nn.embedding_lookup(params=_word_embeddings,ids=self.word_ids)
  word_embeddings_modify = tf.scatter_nd_update(word_embeddings, self.error_word, sum_all)
  

error:
  Tensor conversion requested dtype float32_ref for Tensor with dtype float32

from the error it seems word_embeddings in function scatter_nd_update dtype is tf.float_32,but scatter_nd_update should accepted word_embeddings dtype  tf.float_32_ref ,how can i change word_embeddings's dtype before use tf.scatter_nd_update

**Comment by ppwwyyxx** [comment]
You should not use `scatter_nd_update` in this case because it's for updating variables. You should use `scatter_nd` instead.

**Comment by cjjjy** [comment]
> You should not use `scatter_nd_update` in this case because it's for updating variables. You should use `scatter_nd` instead.

Thanks for your apply ,I just want to update some values in tensor word_embeddings  ,but scatter_nd seems just create a new tensor  by applying sparse updates to individual values or slices within a tensor (initially zero for numeric, empty for string) of the given shape according to indices

**Comment by ppwwyyxx** [comment]
> update some values in tensor word_embeddings 

Your tensor is by definition immutable and you cannot update it.

In your case to avoid the zeros you can do a `scatter_nd` to create a tensor with the differences and add it back to `word_embeddings`.

**Comment by cjjjy** [comment]
Yes ,I understand what you mean ,I googled and found another method it says 
            var_output_2 = tf.Variable(0, dtype=tf.float32,name="a",trainable=False,validate_shape=False)
            var_output_3=tf.assign(var_output_2, word_embeddings, validate_shape=False)
            word_embeddings = tf.scatter_nd_update(var_output_3, self.error_word, sum_all)

**Comment by cjjjy** [comment]
> Yes ,I understand what you mean ,I googled and found another method it says
> var_output_2 = tf.Variable(0, dtype=tf.float32,name="a",trainable=False,validate_shape=False)
> var_output_3=tf.assign(var_output_2, word_embeddings, validate_shape=False)
> word_embeddings = tf.scatter_nd_update(var_output_3, self.error_word, sum_all)



> > update some values in tensor word_embeddings
> 
> Your tensor is by definition immutable and you cannot update it.
> 
> In your case to avoid the zeros you can do a `scatter_nd` to create a tensor with the differences and add it back to `word_embeddings`.

Yes ,I understand what you mean ,I googled and found another method it says
var_output_2 = tf.Variable(0, dtype=tf.float32,name="a",trainable=False,validate_shape=False)
var_output_3=tf.assign(var_output_2, word_embeddings, validate_shape=False)
word_embeddings = tf.scatter_nd_update(var_output_3, self.error_word, sum_all)

**Comment by cjjjy** [comment]
> > update some values in tensor word_embeddings
> 
> Your tensor is by definition immutable and you cannot update it.
> 
> In your case to avoid the zeros you can do a `scatter_nd` to create a tensor with the differences and add it back to `word_embeddings`.

Yes ,I understand what you mean ,I googled and found another method it says
var_output_2 = tf.Variable(0, dtype=tf.float32,name="a",trainable=False,validate_shape=False)
var_output_3=tf.assign(var_output_2, word_embeddings, validate_shape=False)
word_embeddings = tf.scatter_nd_update(var_output_3, self.error_word, sum_all)

**Comment by ppwwyyxx** [comment]
`tf.tensor_scatter_update` seems to be a cleaner method.
